# Pandas

The pandas package is the most important tool at the disposal of Data Scientists and Analysts working in Python
today. The powerful machine learning and glamorous visualization tools may get all the attention, but pandas is the
backbone of most data projects.

This tool is essentially your data’s home. Through pandas, you get acquainted with your data by cleaning,
transforming, and analyzing it.
For example, say you want to explore a dataset stored in a CSV on your computer. Pandas will extract the data from
that CSV into a DataFrame — a table, basically — then let you do things like:

- Calculate statistics and answer questions about the data, like
   - What's the average, median, max, or min of each column?
   - Does column A correlate with column B?
   - What does the distribution of data in column C look like?
- Clean the data by doing things like removing missing values and filtering rows or columns by some criteria
- Visualize the data with help from Matplotlib. Plot bars, lines, histograms, bubbles, and more.
- Store the cleaned, transformed data back into a CSV, other file or database
- Before you jump into the modeling or the complex visualizations you need to have a good understanding of the
- Nature of your dataset and pandas is the best avenue through which to do that.

Once installed, ou just eed to import the `pandas` module to your code

In [1]:
import pandas as pd

# Titanic Dataset

This is a dataset to predict how different variables such as Gender, Class, Age are related with the Survival chances in the sinking of the Titanic, one of the most infamous shipwrecks in history.



| `Variable` | `Defination` | `Key` |
| ------------- |:-------------:| -----:|
|   survival    |   Survival	|   0 = No, 1 = Yes |
|pclass |	Ticket class    |	1 = 1st, 2 = 2nd, 3 = 3rd   |
|sex	|   Sex	    |   |
|Age	|   Age in years    |   |	
|sibsp	|   # of siblings / spouses aboard the Titanic| |	
|parch	|   # of parents / children aboard the Titanic| |	
|ticket	|   Ticket number	|   |
|fare	|   Passenger fare	|   |
|cabin	|   Cabin number    |   |	
|embarked	|   Port of Embarkation	    |   C = Cherbourg, Q = Queenstown, S = Southampton|

Variable Notes

pclass: A proxy for socio-economic status (SES)

1st = Upper

2nd = Middle

3rd = Lower


`age:` Age is fractional if less than 1. If the age is estimated, is it in the form of xx.5


`sibsp:` The dataset defines family relations in this way...

Sibling = brother, sister, stepbrother, stepsister

Spouse = husband, wife (mistresses and fiancés were ignored)


`parch:` The dataset defines family relations in this way...
Parent = mother, father
Child = daughter, son, stepdaughter, stepson
Some children travelled only with a nanny, therefore parch=0 for them.

We'll be using this dataset just to have hands on practice of basic pandas methods and handling missing data

## Reading a `CSV` file in pandas
With CSV files all you need is a single line to load in the data.
CSVs don't have indexes like usual DataFrames, so all we need to do is just designate the `index_col` when reading.

Here we're setting the index to be column zero.

In [2]:
df=pd.read_csv("./raw-data/titanic/train.csv",index_col=0)
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

#### `head()`
The first thing to do when opening a new dataset is print out a few rows to keep as a visual reference. We can accomplish this with `.head()`

`head()` outputs the first five rows of your DataFrame by default, but we could also pass a number as well: `df.head(10)` would output the top ten rows, for example

In [3]:
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

#### `tail()`
To see the last five rows use `.tail()`. It also accepts a number, and in this case we printing the bottom 10 rows:


In [4]:
df.tail(10)

Survived  Pclass                                      Name  \
PassengerId                                                               
882                 0       3                        Markun, Mr. Johann   
883                 0       3              Dahlberg, Miss. Gerda Ulrika   
884                 0       2             Banfield, Mr. Frederick James   
885                 0       3                    Sutehall, Mr. Henry Jr   
886                 0       3      Rice, Mrs. William (Margaret Norton)   
887                 0       2                     Montvila, Rev. Juozas   
888                 1       1              Graham, Miss. Margaret Edith   
889                 0       3  Johnston, Miss. Catherine Helen "Carrie"   
890                 1       1                     Behr, Mr. Karl Howell   
891                 0       3                       Dooley, Mr. Patrick   

                Sex   Age  SibSp  Parch            Ticket     Fare Cabin  \
PassengerId                                                                
882            male  33.0      0      0            349257   7.8958   NaN   
883          female  22.0      0      0              7552  10.5167   NaN   
884            male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN   
885            male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN   
886          female  39.0      0      5            382652  29.1250   NaN   
887            male  27.0      0      0            211536  13.0000   NaN   
888          female  19.0      0      0            112053  30.0000   B42   
889          female   NaN      1      2        W./C. 6607  23.4500   NaN   
890            male  26.0      0      0            111369  30.0000  C148   
891            male  32.0      0      0            370376   7.7500   NaN   

            Embarked  
PassengerId           
882                S  
883                S  
884                S  
885                S  
886                Q  
887                S  
888                S  
889                S  
890                C  
891                Q

#### `shape`
Another fast and useful attribute is `.shape`, which outputs just a tuple of (rows, columns):

Note that `.shape` has no parentheses and is a simple tuple of format (rows, columns).

You'll be going to `.shape` a lot when cleaning and transforming data. For example, you might filter some rows
based on some criteria and then want to know quickly how many rows were removed.


In [5]:
df.shape

(891, 11)

#### `info()`
`.info()` provides the essential details about your dataset, such as the number of rows and columns, the number of
non-null values, what type of data is in each column, and how much memory your DataFrame is using.

Seeing the datatype quickly is actually quite useful. Imagine you just imported some JSON and the integers were
recorded as strings. You go to do some arithmetic and find an "unsupported operand" Exception because you can't
do math with strings. Calling `.info()` will quickly point out that your column you thought was all integers are actually
string objects.


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Name      891 non-null    object 
 3   Sex       891 non-null    object 
 4   Age       714 non-null    float64
 5   SibSp     891 non-null    int64  
 6   Parch     891 non-null    int64  
 7   Ticket    891 non-null    object 
 8   Fare      891 non-null    float64
 9   Cabin     204 non-null    object 
 10  Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 83.5+ KB


#### `columns`
`.columns` come in handy if you want to rename columns by allowing for simple copy and paste, it's
also useful if you need to understand why you are receiving a Key Error when selecting data by column.

But what if we want to lowercase all names? we can do a list comprehension:


In [64]:
print(df.columns)

temp=df
temp.columns=[column.lower() for column in temp.columns]
temp

Index(['survived', 'pclass', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket',
       'fare', 'cabin', 'embarked'],
      dtype='object')


survived  pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          name     sex   age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             sibsp  parch            ticket     fare cabin embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

#### `describe()`
Using `.describe()` on an entire DataFrame we can get a summary of the distribution of continuous variables:

Understanding which numbers are continuous also comes in handy when thinking about the type of plot to use to
represent your data visually.


In [8]:
df.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

`describe()` can also be used on a categorical variable to get the count of rows, unique count of categories, top
category, and freq of top category:

The cell below tells us that the `Embarked` column has 3 unique values, the top value is S, which shows up 644 times (freq).

Whhat we can deduce from that is, out of total 889 passangers, highest number of passangers (644) were onboard from the `Southampton` port

In [9]:
df.Embarked.describe()

count     889
unique      3
top         S
freq      644
Name: Embarked, dtype: object

#### `corr()`
By using the correlation method `.corr()` we can generate the relationship between each continuous variable:

Correlation tables are a numerical representation of the bivariate relationships in the dataset.

Positive numbers indicate a positive correlation — one goes up the other goes up — and negative numbers
represent an inverse correlation — one goes up the other goes down. 1.0 indicates a perfect correlation.


In [10]:
df.corr()

Survived    Pclass       Age     SibSp     Parch      Fare
Survived  1.000000 -0.338481 -0.077221 -0.035322  0.081629  0.257307
Pclass   -0.338481  1.000000 -0.369226  0.083081  0.018443 -0.549500
Age      -0.077221 -0.369226  1.000000 -0.308247 -0.189119  0.096067
SibSp    -0.035322  0.083081 -0.308247  1.000000  0.414838  0.159651
Parch     0.081629  0.018443 -0.189119  0.414838  1.000000  0.216225
Fare      0.257307 -0.549500  0.096067  0.159651  0.216225  1.000000

#### `Inferences from above corelation`
- `Pclass` is significantly related to `survival`, which means higher the `VALUE of class`, lower is their survival. Which makes sense because, 1st class people had higher chances of survival then that of 3rd class people knowing that officials were more eager to save 1st class people.

- We get one more confirmation by relating `Fare` and `Survived`. Higher the fare higher are their survival chances. The people who pay higher are 1st class and 2nd class people, and their survival chances are higher as per the correlation. Hence this is a confirmation to the corollary we stated before.

- Another very trivial deduction is, people in 1st class have paid more than those of 3rd class [See relation between `Fare` & `Pclass`]

# `isnull()`
`.isnull()` returns a DataFrame where each cell is either True or False depending on that cell's null status.
To count the number of nulls in each column we use an aggregate function for summing.

`.isnull()` just by iteself isn't very useful, and is usually used in conjunction with other methods, like sum().
We can see now that our data has 177 missing values for Age and 687 missing values for Cabin and 2 missing values for Embarked column.


In [11]:
df.isnull().sum()

Survived      0
Pclass        0
Name          0
Sex           0
Age         177
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

# Dealing with Missing Values

## Three Approaches


### 1) A Simple Option: Drop Columns with Missing Values

The simplest option is to drop columns with missing values. 

![tut2_approach1](https://i.imgur.com/Sax80za.png)

Unless most values in the dropped columns are missing, the model loses access to a lot of (potentially useful!) information with this approach.  As an extreme example, consider a dataset with 10,000 rows, where one important column is missing a single entry. This approach would drop the column entirely!

### 2) A Better Option: Imputation

**Imputation** fills in the missing values with some number.  For instance, we can fill in the mean value along each column. 

![tut2_approach2](https://i.imgur.com/4BpnlPA.png)

The imputed value won't be exactly right in most cases, but it usually leads to more accurate models than you would get from dropping the column entirely.

### 3) An Extension To Imputation

Imputation is the standard approach, and it usually works well. However, imputed values may be systematically above or below their actual values (which weren't collected in the dataset). Or rows with missing values may be unique in some other way. In that case, your model would make better predictions by considering which values were originally missing.  

![tut3_approach3](https://i.imgur.com/UWOyg4a.png)

In this approach, we impute the missing values, as before.  And, additionally, for each column with missing entries in the original dataset, we add a new column that shows the location of the imputed entries.

In some cases, this will meaningfully improve results. In other cases, it doesn't help at all.

# Example 

In the example, we will work with the [Titanic dataset](https://www.kaggle.com/c/titanic/data).


Below is the column-wise percentage missing data in our data-frame.

`Cabin` has highest (more than 3/4 th) amount of missing data while `Embarked` has lowest (only 2 values).

In [12]:
# % missing in Age column
(df.Age.isnull().sum()/df.shape[0])*100

19.865319865319865

In [13]:
# % missing in Cabin Column
(df.Cabin.isnull().sum()/df.shape[0])*100

77.10437710437711

In [14]:
# % missing in Embarked column
(df.Embarked.isnull().sum()/df.shape[0])*100

0.22446689113355783

Now we replace all nulls values in the dataset with different values [column Age with mean, column Cabin with 'B98' and column Embarked with 'S'] and make a whole new dataframe if it named `data`.

Note that our original dataframe `df` is untouched.

In [15]:
fill_missing={'Age':df.Age.mean(),'Cabin':'B98','Embarked':'S'}
data=df.fillna(value=fill_missing)
data.isnull().sum()

Survived    0
Pclass      0
Name        0
Sex         0
Age         0
SibSp       0
Parch       0
Ticket      0
Fare        0
Cabin       0
Embarked    0
dtype: int64

Now lets try and use SimpleImputer class from sklearn.impute module and fill our missing values.

Now at this point, we should know that `SimpleImputer class fails to treat String datatypes`. So we will have to get rid of all the String datatypes from our dataframe. We'll do that by simple slicing and selecting srea of interest. We will keep all the features in a dataframe named `X` and we'll have the `Survived` i.e., values to be predicted in `y` dataframe.

In [55]:
# Selecting only the numeric data
X=df.select_dtypes(exclude=['object'])
y=df.Survived
X.isnull().sum()

Survived      0
Pclass        0
Age         177
SibSp         0
Parch         0
Fare          0
dtype: int64

In [56]:
from sklearn.impute import SimpleImputer

# Imputation
my_imputer = SimpleImputer()
imputed_X = pd.DataFrame(my_imputer.fit_transform(X))


imputed_X.isnull().sum()

0    0
1    0
2    0
3    0
4    0
5    0
dtype: int64

In [60]:
imputed_X.head(3)

Survived  Pclass   Age  SibSp  Parch     Fare
0       0.0     3.0  22.0    1.0    0.0   7.2500
1       1.0     1.0  38.0    1.0    0.0  71.2833
2       1.0     3.0  26.0    0.0    0.0   7.9250

We can see that the SimpleImputer removed all the column names. Let's put them back just to have a uniformity

In [61]:
imputed_X.columns=X.columns
imputed_X.head()

Survived  Pclass   Age  SibSp  Parch     Fare
0       0.0     3.0  22.0    1.0    0.0   7.2500
1       1.0     1.0  38.0    1.0    0.0  71.2833
2       1.0     3.0  26.0    0.0    0.0   7.9250
3       1.0     1.0  35.0    1.0    0.0  53.1000
4       0.0     3.0  35.0    0.0    0.0   8.0500